src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -183,126 +319,191 @@ number of sites `M` inferred from the number of modes in `address`.
-        address, t_vec, u_mat, v_mat, pot_vec, geometry,
+    return HubbardRealSpace{
+        TT,
+        C,
+        typeof(address),
+        typeof(geometry),
+        D,
+        typeof(t_mat),
+        typeof(u_mat),
+        typeof(w_mat),
+        typeof(v_mat),
+        typeof(pot_vec),
+    }(
+        address, t_mat, u_mat, w_mat, v_mat, pot_vec, geometry
     )
 end
 
+# Convert input of t or v to static matrix
+function _t_or_v_to_matrix(name, value, num_comps, num_dims; zero_is_nothing=true)
+    if zero_is_nothing && iszero(value)
+        return nothing
+    else
+        if value isa Number
+            value = fill(value, (num_comps, num_dims))
+        elseif size(value, 2) == 1 # column-vector
+            value = reduce(hcat, value for _ in 1:num_dims)
+        elseif size(value, 1) == 1 # row-vector
+            value = reduce(vcat, value for _ in 1:num_comps)
+        end
+        if (size(value, 1), size(value, 2)) ≠ (num_comps, num_dims)
+            throw(
+                ArgumentError(
+                    "`$name` must be a number, $num_comps × $num_dims matrix, or vector of length $num_comps",
+                ),
+            )
+        end
+        return SMatrix{num_comps,num_dims,float(eltype(value))}(value)
+    end
+end
+# Convert input of u or w to static matrix
+function _u_or_w_to_matrix(name, value, num_comps)
+    if iszero(value)
+        return nothing
+    else
+        if value isa Number
+            value = fill(value, (num_comps, num_comps))
+        end
+        if (size(value, 1), size(value, 2)) ≠ (num_comps, num_comps) ||
+            value isa Matrix && !issymmetric(value)
+            throw(
+                ArgumentError(
+                    "`$name` must be a number or a symmetric $num_comps × $num_comps matrix"
+                ),
+            )
+        end
+        return SMatrix{num_comps,num_comps,Float64}(value)
+    end
+end
+
 """
     warn_fermi_interaction(address, u)
 
-Warn if interaction matrix `u` does not make sense for `address`.
+Warn and modify if interaction matrix `u` does not make sense for `address`.
 """
 function warn_fermi_interaction(address::CompositeFS, u)
     C = num_components(address)
+    U = Matrix(u)
     for c in 1:C
-        if address.components[c] isa FermiFS && u ≠ ones(C,C) && u[c,c] ≠ 0
+        if address.components[c] isa FermiFS && u[c, c] ≠ 0
             @warn "component $(c) is fermionic, but was given a self-interaction " *
                 "strength of $(u[c,c])" maxlog=1
+            U[c, c] = 0
         end
     end
+    return SMatrix{size(U)...,eltype(u)}(U)
 end
 function warn_fermi_interaction(address::FermiFS, u)
-    if u ≠ ones(1, 1) && u[1, 1] ≠ 0
+    if u[1, 1] ≠ 0
         @warn "address is fermionic, but was given a self-interaction " *
             "strength of $(u[1,1])" maxlog=1
     end
+    return SMatrix{1,1,eltype(u)}(0)
 end
```

Comment:
I just noticed the tests are failing. They started failing when this was changed. I'd suggest reverting it back, setting `U[c, c]` to zero doesn't do anything here (since the fermions don't interact either way), and the `u ≠ ones(C,C)` is there to not trigger the warning when using the default interaction of 1.